File: src/parsing/functions/temporal_utils.ts
/**
 * Shared utility functions for temporal (date/time) operations.
 *
 * These helpers are used by the datetime, date, time, localdatetime,
 * localtime, and timestamp functions.
 */

/**
 * Computes the ISO day of the week (1 = Monday, 7 = Sunday) matching Neo4j convention.
 */
function isoDayOfWeek(d: Date): number {
    const jsDay = d.getDay(); // 0 = Sunday, 6 = Saturday
    return jsDay === 0 ? 7 : jsDay;
}

/**
 * Computes the day of the year (1-based).
 */
function dayOfYear(d: Date): number {
    const start = new Date(d.getFullYear(), 0, 0);
    const diff = d.getTime() - start.getTime();
    const oneDay = 1000 * 60 * 60 * 24;
    return Math.floor(diff / oneDay);
}

/**
 * Computes the quarter (1-4) from a month (0-11).
 */
function quarter(month: number): number {
    return Math.floor(month / 3) + 1;
}

/**
 * Pads a number to a given width with leading zeros.
 */
function pad(n: number, width: number = 2): string {
    return String(n).padStart(width, "0");
}

/**
 * Formats a timezone offset in ±HH:MM format.
 */
function formatTimezoneOffset(d: Date): string {
    const offset = -d.getTimezoneOffset();
    if (offset === 0) return "Z";
    const sign = offset > 0 ? "+" : "-";
    const absOffset = Math.abs(offset);
    const hours = Math.floor(absOffset / 60);
    const minutes = absOffset % 60;
    return `${sign}${pad(hours)}:${pad(minutes)}`;
}

/**
 * Parses a temporal argument (string or map) into a Date object.
 *
 * @param arg - The argument to parse (string or object with year/month/day/hour/minute/second/millisecond)
 * @param fnName - The calling function name for error messages
 * @returns A Date object
 */
export function parseTemporalArg(arg: any, fnName: string): Date {
    if (typeof arg === "string") {
        const d = new Date(arg);
        if (isNaN(d.getTime())) {
            throw new Error(`${fnName}(): Invalid temporal string: '${arg}'`);
        }
        return d;
    }

    if (typeof arg === "number") {
        // Treat as epoch milliseconds
        return new Date(arg);
    }

    if (typeof arg === "object" && arg !== null && !Array.isArray(arg)) {
        // Map-style construction: {year, month, day, hour, minute, second, millisecond}
        const year = arg.year ?? new Date().getFullYear();
        const month = (arg.month ?? 1) - 1; // JS months are 0-based
        const day = arg.day ?? 1;
        const hour = arg.hour ?? 0;
        const minute = arg.minute ?? 0;
        const second = arg.second ?? 0;
        const millisecond = arg.millisecond ?? 0;
        return new Date(year, month, day, hour, minute, second, millisecond);
    }

    throw new Error(
        `${fnName}(): Expected a string, number (epoch millis), or map argument, got ${typeof arg}`
    );
}

/**
 * Builds a datetime result object with full temporal properties.
 *
 * @param d - The Date object
 * @param utc - If true, use UTC values; if false, use local values
 * @returns An object with year, month, day, hour, minute, second, millisecond,
 *          epochMillis, epochSeconds, dayOfWeek, dayOfYear, quarter, formatted
 */
export function buildDatetimeObject(d: Date, utc: boolean): Record<string, any> {
    const year = utc ? d.getUTCFullYear() : d.getFullYear();
    const month = utc ? d.getUTCMonth() + 1 : d.getMonth() + 1;
    const day = utc ? d.getUTCDate() : d.getDate();
    const hour = utc ? d.getUTCHours() : d.getHours();
    const minute = utc ? d.getUTCMinutes() : d.getMinutes();
    const second = utc ? d.getUTCSeconds() : d.getSeconds();
    const millisecond = utc ? d.getUTCMilliseconds() : d.getMilliseconds();

    const datePart = new Date(year, month - 1, day);

    const formatted = utc
        ? d.toISOString()
        : `${year}-${pad(month)}-${pad(day)}T${pad(hour)}:${pad(minute)}:${pad(second)}.${pad(millisecond, 3)}`;

    return {
        year,
        month,
        day,
        hour,
        minute,
        second,
        millisecond,
        epochMillis: d.getTime(),
        epochSeconds: Math.floor(d.getTime() / 1000),
        dayOfWeek: isoDayOfWeek(datePart),
        dayOfYear: dayOfYear(datePart),
        quarter: quarter(month - 1),
        formatted,
    };
}

/**
 * Builds a date result object (no time component).
 *
 * @param d - The Date object
 * @returns An object with year, month, day, epochMillis, dayOfWeek, dayOfYear, quarter, formatted
 */
export function buildDateObject(d: Date): Record<string, any> {
    const year = d.getFullYear();
    const month = d.getMonth() + 1;
    const day = d.getDate();

    // Strip time component for epoch calculation
    const dateOnly = new Date(year, month - 1, day);

    return {
        year,
        month,
        day,
        epochMillis: dateOnly.getTime(),
        dayOfWeek: isoDayOfWeek(dateOnly),
        dayOfYear: dayOfYear(dateOnly),
        quarter: quarter(month - 1),
        formatted: `${year}-${pad(month)}-${pad(day)}`,
    };
}

/**
 * Builds a time result object (no date component).
 *
 * @param d - The Date object
 * @param utc - If true, use UTC values; if false, use local values
 * @returns An object with hour, minute, second, millisecond, formatted
 */
export function buildTimeObject(d: Date, utc: boolean): Record<string, any> {
    const hour = utc ? d.getUTCHours() : d.getHours();
    const minute = utc ? d.getUTCMinutes() : d.getMinutes();
    const second = utc ? d.getUTCSeconds() : d.getSeconds();
    const millisecond = utc ? d.getUTCMilliseconds() : d.getMilliseconds();

    const timePart = `${pad(hour)}:${pad(minute)}:${pad(second)}.${pad(millisecond, 3)}`;
    const formatted = utc ? `${timePart}Z` : timePart;

    return {
        hour,
        minute,
        second,
        millisecond,
        formatted,
    };
}

File: flowquery-py/src/parsing/functions/temporal_utils.py
"""Shared utility functions for temporal (date/time) operations.

These helpers are used by the datetime_, date_, time_, localdatetime,
localtime, and timestamp functions.
"""

from datetime import date, datetime, timezone
from typing import Any, Dict


def iso_day_of_week(d: date) -> int:
    """Computes the ISO day of the week (1 = Monday, 7 = Sunday)."""
    return d.isoweekday()


def day_of_year(d: date) -> int:
    """Computes the day of the year (1-based)."""
    return d.timetuple().tm_yday


def quarter(month: int) -> int:
    """Computes the quarter (1-4) from a month (1-12)."""
    return (month - 1) // 3 + 1


def parse_temporal_arg(arg: Any, fn_name: str) -> datetime:
    """Parses a temporal argument (string, number, or map) into a datetime object.

    Args:
        arg: The argument to parse (string, number, or dict with components)
        fn_name: The calling function name for error messages

    Returns:
        A datetime object
    """
    if isinstance(arg, str):
        try:
            return datetime.fromisoformat(arg.replace("Z", "+00:00"))
        except ValueError:
            raise ValueError(f"{fn_name}(): Invalid temporal string: '{arg}'")

    if isinstance(arg, (int, float)):
        # Treat as epoch milliseconds
        return datetime.fromtimestamp(arg / 1000, tz=timezone.utc)

    if isinstance(arg, dict):
        # Map-style construction: {year, month, day, hour, minute, second, millisecond}
        now = datetime.now()
        year = arg.get("year", now.year)
        month = arg.get("month", 1)
        day = arg.get("day", 1)
        hour = arg.get("hour", 0)
        minute = arg.get("minute", 0)
        second = arg.get("second", 0)
        millisecond = arg.get("millisecond", 0)
        return datetime(year, month, day, hour, minute, second, millisecond * 1000)

    raise ValueError(
        f"{fn_name}(): Expected a string, number (epoch millis), or map argument, "
        f"got {type(arg).__name__}"
    )


def build_datetime_object(d: datetime, utc: bool) -> Dict[str, Any]:
    """Builds a datetime result object with full temporal properties.

    Args:
        d: The datetime object
        utc: If True, use UTC values; if False, use local values

    Returns:
        A dict with year, month, day, hour, minute, second, millisecond,
        epochMillis, epochSeconds, dayOfWeek, dayOfYear, quarter, formatted
    """
    if utc:
        if d.tzinfo is None:
            d = d.replace(tzinfo=timezone.utc)
        d_utc = d.astimezone(timezone.utc)
        year = d_utc.year
        month = d_utc.month
        day = d_utc.day
        hour = d_utc.hour
        minute = d_utc.minute
        second = d_utc.second
        millisecond = d_utc.microsecond // 1000
        formatted = d_utc.strftime("%Y-%m-%dT%H:%M:%S.") + f"{millisecond:03d}Z"
    else:
        if d.tzinfo is not None:
            d = d.astimezone(tz=None).replace(tzinfo=None)
        year = d.year
        month = d.month
        day = d.day
        hour = d.hour
        minute = d.minute
        second = d.second
        millisecond = d.microsecond // 1000
        formatted = d.strftime("%Y-%m-%dT%H:%M:%S.") + f"{millisecond:03d}"

    date_part = date(year, month, day)
    epoch_millis = int(d.timestamp() * 1000) if d.tzinfo else int(
        datetime(year, month, day, hour, minute, second, millisecond * 1000).timestamp() * 1000
    )

    return {
        "year": year,
        "month": month,
        "day": day,
        "hour": hour,
        "minute": minute,
        "second": second,
        "millisecond": millisecond,
        "epochMillis": epoch_millis,
        "epochSeconds": epoch_millis // 1000,
        "dayOfWeek": iso_day_of_week(date_part),
        "dayOfYear": day_of_year(date_part),
        "quarter": quarter(month),
        "formatted": formatted,
    }


def build_date_object(d: datetime) -> Dict[str, Any]:
    """Builds a date result object (no time component).

    Args:
        d: The datetime object

    Returns:
        A dict with year, month, day, epochMillis, dayOfWeek, dayOfYear, quarter, formatted
    """
    year = d.year
    month = d.month
    day_val = d.day

    date_only = datetime(year, month, day_val)
    epoch_millis = int(date_only.timestamp() * 1000)

    date_part = date(year, month, day_val)

    return {
        "year": year,
        "month": month,
        "day": day_val,
        "epochMillis": epoch_millis,
        "dayOfWeek": iso_day_of_week(date_part),
        "dayOfYear": day_of_year(date_part),
        "quarter": quarter(month),
        "formatted": f"{year}-{month:02d}-{day_val:02d}",
    }


def build_time_object(d: datetime, utc: bool) -> Dict[str, Any]:
    """Builds a time result object (no date component).

    Args:
        d: The datetime object
        utc: If True, use UTC values; if False, use local values

    Returns:
        A dict with hour, minute, second, millisecond, formatted
    """
    if utc:
        if d.tzinfo is None:
            d = d.replace(tzinfo=timezone.utc)
        d_utc = d.astimezone(timezone.utc)
        hour = d_utc.hour
        minute = d_utc.minute
        second = d_utc.second
        millisecond = d_utc.microsecond // 1000
    else:
        if d.tzinfo is not None:
            d = d.astimezone(tz=None).replace(tzinfo=None)
        hour = d.hour
        minute = d.minute
        second = d.second
        millisecond = d.microsecond // 1000

    time_part = f"{hour:02d}:{minute:02d}:{second:02d}.{millisecond:03d}"
    formatted = f"{time_part}Z" if utc else time_part

    return {
        "hour": hour,
        "minute": minute,
        "second": second,
        "millisecond": millisecond,
        "formatted": formatted,
    }
